csharp/src/iceboxnet/ServiceManager.cs

```diff
@@ -5,192 +5,142 @@
 
         internal ServiceManager(Communicator communicator, string[] args)
         {
             _communicator = communicator;
             _logger = _communicator.Logger;
 
-            _adminEnabled = _communicator.GetPropertyAsBool("Ice.Admin.Enabled") ??
+            bool adminEnabled = _communicator.GetPropertyAsBool("Ice.Admin.Enabled") ??
                 _communicator.GetProperty("Ice.Admin.Endpoints") != null;
 
-            if (_adminEnabled)
+            if (adminEnabled)
             {
-                _adminFacetFilter = new HashSet<string>(_communicator.GetPropertyAsList("Ice.Admin.Facets") ??
-                                                        Array.Empty<string>());
+                _adminFacetFilter = new HashSet<string>(
+                    _communicator.GetPropertyAsList("Ice.Admin.Facets") ?? Array.Empty<string>());
             }
 
             _argv = args;
             _traceServiceObserver = _communicator.GetPropertyAsInt("IceBox.Trace.ServiceObserver") ?? 0;
         }
 
-        public void AddObserver(IServiceObserverPrx? observer, Current current, CancellationToken cancel)
+        public ValueTask AddObserverAsync(IServiceObserverPrx observer, Current current, CancellationToken cancel)
         {
-            // Null observers and duplicate registrations are ignored
-            if (observer != null)
+            lock (_mutex)
             {
-                string[] activeServices;
-                lock (_mutex)
+                try
                 {
-                    try
-                    {
-                        _observers.Add(observer);
-                    }
-                    catch (ArgumentException)
-                    {
-                        return;
-                    }
+                    _observers.Add(observer);
+                }
+                catch (ArgumentException)
+                {
+                    return default; // ignore duplicates.
+                }
+            }
 
-                    if (_traceServiceObserver >= 1)
-                    {
-                        _logger.Trace("IceBox.ServiceObserver", $"Added service observer {observer}");
-                    }
+            if (_traceServiceObserver >= 1)
+            {
+                _logger.Trace("IceBox.ServiceObserver", $"added service observer {observer}");
+            }
 
-                    activeServices = _services.Where(info => info.Status == ServiceStatus.Started).Select(
-                        info => info.Name).ToArray();
-                }
+            IEnumerable<string> activeServices = _services.Where(info => info.IsActive).Select(info => info.Name);
+
+            if (activeServices.Any())
+            {
+                _ = StartedAsync(observer, activeServices);
+            }
+            return default;
 
-                if (activeServices.Length > 0)
+            async Task StartedAsync(IServiceObserverPrx observer, IEnumerable<string> services)
```

Comment:
What is the advantage of calling the observer from this local function, can we just make AddObserverAsync async and await the call to ` observer.ServicesStartedAsync` from the dispatch?

Reply:
`AddObserverAsync` notifies the observer but does not wait for the observer's response. How would you do this without a local function?

Reply:
I was thinking that we can wait from the dispatch, as all this will be async, why does it matter if the Dispatch terminates after `AddObserverAsync` completes

Reply:
I am not following: `AddObserverAsync` is the servant dispatch method. The operation completes only when this dispatch completes. How would you rewrite AddObserverAsync?

Reply:
I mean make `AddObserverAsync` async and just do
```
if (activeServices.Any())
{
    try
    {
        // We don't want to forward the cancel of a dispatch that has already completed.
        await observer.ServicesStartedAsync(services, cancel: default);
    }
    catch (Exception ex)
    {
        RemoveObserver(observer, ex);
    }
}
```

Reply:
But then the addObserver operation would not complete until the observers notification is complete, which is what we want to avoid here.